src/core/scalar.rs

```diff
@@ -0,0 +1,312 @@
+) -> bool {
+    val.as_i64().map_or(false, |n| fxn(n).is_ok())
+}
+
+fn eval_unsigned<
+    'a,
+    Num,
+    Value: JsonLike<'a> + 'a,
+    F: Fn(u64) -> Result<Num, std::num::TryFromIntError>,
+>(
+    val: &'a Value,
+    fxn: F,
+) -> bool {
+    val.as_u64().map_or(false, |n| fxn(n).is_ok())
+}
+
+impl ScalarType {
+    pub fn validate<'a, Value: JsonLike<'a> + 'a>(&self, value: &'a Value) -> bool {
+        match self {
+            ScalarType::JSON => true,
+            ScalarType::Empty => true,
+            ScalarType::Email => eval_str(value, |s| {
+                async_graphql::validators::email(&s.to_string()).is_ok()
+            }),
+            ScalarType::PhoneNumber => eval_str(value, |s| phonenumber::parse(None, s).is_ok()),
+            ScalarType::Date => {
+                eval_str(value, |s| chrono::DateTime::parse_from_rfc3339(s).is_ok())
+            }
+            ScalarType::Url => eval_str(value, |s| url::Url::parse(s).is_ok()),
+            ScalarType::Bytes => value.as_str().is_some(),
+
+            ScalarType::Int64 => eval_str(value, |s| s.parse::<i64>().is_ok()),
+            ScalarType::UInt64 => eval_str(value, |s| s.parse::<u64>().is_ok()),
+            ScalarType::Int128 => eval_str(value, |s| s.parse::<i128>().is_ok()),
+            ScalarType::UInt128 => eval_str(value, |s| s.parse::<u128>().is_ok()),
+
+            ScalarType::Int8 => eval_signed(value, i8::try_from),
+            ScalarType::Int16 => eval_signed(value, i16::try_from),
+            ScalarType::Int32 => eval_signed(value, i32::try_from),
+
+            ScalarType::UInt8 => eval_unsigned(value, u8::try_from),
+            ScalarType::UInt16 => eval_unsigned(value, u16::try_from),
+            ScalarType::UInt32 => eval_unsigned(value, u32::try_from),
+        }
+    }
+    pub fn get_scalar(name: &str) -> ScalarType {
+        CUSTOM_SCALARS.get(name).cloned().unwrap_or(Self::Empty)
+    }
+    pub fn name(&self) -> String {
+        self.to_string()
+    }
+    pub fn scalar_definition(&self) -> async_graphql::parser::types::TypeSystemDefinition {
+        let schemars = self.schema();
+        tailcall_typedefs_common::scalar_definition::into_scalar_definition(schemars, &self.name())
+    }
+    fn schema_inner(&self, type_of: InstanceType, description: &str) -> Schema {
+        let format = match type_of {
+            InstanceType::Integer => Some(self.name().to_lowercase()),
+            _ => None,
+        };
+        let mut value = serde_json::json!(
+            {
+                "title": self.name(),
+                "type": type_of,
+                "description": description,
+            }
+        );
+        if let Some(format) = format {
+            value["format"] = serde_json::json!(format);
+        }
+
+        let metadata = serde_json::from_value(value).unwrap();
+        Schema::Object(SchemaObject { metadata: Some(Box::new(metadata)), ..Default::default() })
+    }
+    pub fn schema(&self) -> Schema {
```

Comment:
Can't we get the schema using schemar for each enum variant?

Reply:
Introduced a Doc macro which gets doc comments from varients' instance. We can not do this using schemars